### PR TITLE
fix(cli): snapshot session usage during long runs

### PR DESCRIPTION
**Pull request by tt-a1i**
## What this PR does

This PR keeps the existing final session-usage flush, and also writes a best-effort usage snapshot every 5 minutes while the CLI process is still running. Snapshots are only written after the session has recorded at least one API request, and the timer is unref'ed so it cannot keep the process alive.

The snapshot path reuses the same session summary writer as shutdown. Existing usage-history loading already keeps the latest record per session ID, so periodic snapshots update the visible session summary without inflating aggregate totals.

## Why it's needed

Long-running or unhealthy sessions can accumulate meaningful token usage before they exit cleanly. Today `usage_record.jsonl` only receives the session summary during cleanup, so a process that keeps running for hours can leave no partial session summary for `/stats` or the dashboard until it finally shuts down.

Periodic snapshots close that reporting blind spot for active sessions. This does not try to solve process-level watchdogs or automatic runtime cutoff; it only makes the usage history less dependent on a clean exit.

### How to verify

Reviewers can confirm that sessions with no API requests do not write snapshots, sessions with API activity write the current session ID/project/start/end/metrics, and the timer performs both interval snapshots and a final cleanup flush. Also confirm that existing usage-history dedup still keeps one aggregate record per session ID.

Commands run locally:

```bash
cd packages/cli && npx vitest run src/gemini.test.tsx --coverage=false
cd packages/core && npx vitest run src/services/usageHistoryService.test.ts --coverage=false
npm exec eslint -- packages/cli/src/gemini.tsx packages/cli/src/gemini.test.tsx
npx prettier --check packages/cli/src/gemini.tsx packages/cli/src/gemini.test.tsx
npm run build --workspace=packages/cli
git diff --check
```

### Evidence (Before & After)

Before: the CLI registered a single best-effort session usage write during exit cleanup. If a long-running session did not exit cleanly, `usage_record.jsonl` could have no partial summary for that session.

After: active sessions write a best-effort snapshot every 5 minutes and still write a final snapshot during cleanup. Existing last-wins session dedup keeps periodic snapshot records from double-counting aggregate usage.

### Tested on

|     OS     | Status |
| :--------: | :----: |
|  🍏 macOS  | ✅ tested |
| 🪟 Windows | ⚠️ not tested |
|  🐧 Linux  | ⚠️ not tested |

### Environment (optional)

Local macOS checkout with repository npm dependencies installed. Validation is via focused unit tests, touched-file lint/format checks, and the CLI package build.

## Risk & Scope

- Main risk or tradeoff: active long-running sessions append one summary line every 5 minutes, so `usage_record.jsonl` can grow by about 12 lines per active hour. Loading already dedups by session ID before aggregation.
- Not validated / out of scope: process-level watchdogs, max runtime enforcement, and cases where the Node event loop is fully blocked and cannot run timers.
- Breaking changes / migration notes: none.

## Linked Issues

Refs #5964

## AI Assistance Disclosure

I used Codex to review the changes, sanity-check the implementation against existing patterns, and help spot potential edge cases.

<details>
<summary>中文说明</summary>

## 这个 PR 做了什么

这个 PR 保留现有的会话结束时 usage flush，同时在 CLI 进程仍在运行时每 5 分钟写入一次 best-effort usage snapshot。只有当会话已经记录过至少一次 API 请求时才会写 snapshot，并且 timer 使用 `unref()`，不会阻止进程退出。

snapshot 路径复用和 shutdown 相同的会话汇总写入逻辑。现有 usage history 读取逻辑已经按 session ID 保留最新记录，所以周期性 snapshot 能更新可见的会话汇总，同时不会放大聚合统计。

## 为什么需要

长时间运行或异常的会话，可能在干净退出之前已经产生了明显 token 用量。当前 `usage_record.jsonl` 只有在 cleanup 时才会收到会话汇总，所以一个持续运行数小时的进程，在最终退出前可能不会给 `/stats` 或 dashboard 留下任何部分会话汇总。

周期性 snapshot 能补上活跃会话的这个统计盲点。这个 PR 不尝试解决进程级 watchdog 或自动运行时切断；它只让 usage history 不再完全依赖干净退出。

### 如何验证

reviewer 可以确认：没有 API 请求的会话不会写 snapshot；有 API 活动的会话会写入当前 session ID、project、start/end time 和 metrics；timer 会执行 interval snapshot，并在 cleanup 时再执行 final flush。同时确认现有 usage-history dedup 仍然保证每个 session ID 在聚合里只算一条记录。

本地运行的命令：

```bash
cd packages/cli && npx vitest run src/gemini.test.tsx --coverage=false
cd packages/core && npx vitest run src/services/usageHistoryService.test.ts --coverage=false
npm exec eslint -- packages/cli/src/gemini.tsx packages/cli/src/gemini.test.tsx
npx prettier --check packages/cli/src/gemini.tsx packages/cli/src/gemini.test.tsx
npm run build --workspace=packages/cli
git diff --check
```

### 证据（修改前 / 修改后）

修改前：CLI 只在 exit cleanup 中注册了一次 best-effort 会话 usage 写入。如果长时间运行的会话没有干净退出，`usage_record.jsonl` 可能完全没有该会话的部分汇总。

修改后：活跃会话每 5 分钟写入一次 best-effort snapshot，并且仍然在 cleanup 时写最终 snapshot。现有 last-wins session dedup 会避免周期性 snapshot 记录在聚合 usage 时重复计数。

### 测试环境

|     OS     | Status |
| :--------: | :----: |
|  🍏 macOS  | ✅ tested |
| 🪟 Windows | ⚠️ not tested |
|  🐧 Linux  | ⚠️ not tested |

### 环境（可选）

本地 macOS checkout，已安装仓库 npm 依赖。验证方式包括聚焦单测、触达文件 lint/format 检查，以及 CLI package build。

## 风险与范围

- 主要风险或权衡：活跃长会话每 5 分钟会追加一条 summary line，所以 `usage_record.jsonl` 每个活跃小时大约会增加 12 行。读取时已经先按 session ID 去重再聚合。
- 未验证 / 不在范围内：进程级 watchdog、最大运行时强制限制，以及 Node event loop 完全阻塞导致 timer 无法执行的情况。
- 破坏性变更 / 迁移说明：无。

## 关联 Issue

Refs #5964

## AI 协助披露

I used Codex to review the changes, sanity-check the implementation against existing patterns, and help spot potential edge cases.

</details>

### CODE REVIEW

**Comment by tt-a1i** [comment]
Merged current upstream/main (22bacfe249) and resolved the gemini.test.tsx import conflict. I also simplified the timer test seam to use Vitest fake timers and the existing cleanup mock.

Local checks are green:
- CLI gemini tests: 79 passed
- usageHistoryService tests: 30 passed
- CLI build + typecheck
- touched-file Prettier + ESLint
- git diff --check

File: packages/cli/src/gemini.tsx [hunk 1]
```diff
@@ -261,6 +263,65 @@ function getSignalExitCode(signal: NodeJS.Signals): number {
...
+  config: Pick<Config, 'getProjectRoot' | 'getSessionId'>,
+  deps?: Partial<SessionUsageSnapshotDeps>,
+): void {
+  const snapshotDeps: SessionUsageSnapshotDeps = {
+    getMetrics: deps?.getMetrics ?? (() => uiTelemetryService.getMetrics()),
+    getSessionStartTime:
+      deps?.getSessionStartTime ??
+      (() => uiTelemetryService.getSessionStartTime()),
+    persist: deps?.persist ?? persistSessionUsage,
+    now: deps?.now ?? (() => new Date()),
```
[the remark this answers is not in the document]
Comment (tt-a1i) [reply packages/cli/src/gemini.tsx]: moved the defaults into one shared defaultSnapshotDeps source and resolve overrides from it for both direct and periodic flushes.

File: packages/cli/src/gemini.tsx [hunk 2]
```diff
@@ -261,6 +263,65 @@ function getSignalExitCode(signal: NodeJS.Signals): number {
...
+  });
+  return true;
+}
+
+export function startSessionUsageSnapshots(
+  config: Pick<Config, 'getProjectRoot' | 'getSessionId'>,
+  deps?: Partial<SessionUsageSnapshotDeps>,
+): void {
+  const snapshotDeps: SessionUsageSnapshotDeps = {
+    getMetrics: deps?.getMetrics ?? (() => uiTelemetryService.getMetrics()),
```
[the remark this answers is not in the document]
Comment (tt-a1i) [reply packages/cli/src/gemini.tsx]: switched snapshots to per-session metrics and per-session start times, and flush the outgoing session before /clear, /resume, and /branch transitions. Fork-inherited telemetry is excluded from the branch's own usage, including transcript recovery.

File: packages/cli/src/gemini.tsx [hunk 3]
```diff
@@ -949,28 +1010,9 @@ export async function main() {
...
-          project: config.getProjectRoot(),
-          metrics,
-        });
-      } catch {
-        // Best-effort — don't block shutdown
-      }
-    });
+    // Persist session usage periodically and on shutdown. Must register before
+    // config.shutdown(), which clears telemetry state.
+    startSessionUsageSnapshots(config);
```
[the remark this answers is not in the document]
Comment (tt-a1i) [reply packages/cli/src/gemini.tsx:967]: fixed by marking periodic writes as partial. Readers now replay partial sessions and choose the newer transcript tail, deletion salvage finalizes partial records, and a newer post-resume partial is no longer hidden by an older final. Added live-load and deletion regression tests.

**Review by doudouOUC — COMMENTED** [review]
Traced the snapshot lifecycle and the per-session attribution at the PR head. **No blocking issues found.**

- Double counting is impossible by construction: every snapshot persists the *cumulative* session metrics under the session id, and the read side dedupes by `sessionId` (last-wins, final-beats-partial). `persistUsageBeforeTranscriptDeletion` skips when a final record exists and otherwise finalizes the partial from whichever source has the newer tail; `mergePartialRecords` lets a newer rebuilt record supersede a stale partial and never touches final ones. So the 5-minute partials, the `/clear` flush, the shutdown flush and the delete-time salvage converge to one record per session.
- The fingerprint gate on the periodic flush (`JSON.stringify(metrics)` unchanged → skip) keeps the 5-minute timer quiet on an idle session, and the timer is `unref`'d so it cannot hold the process open; the cleanup flush runs `isPartial: false` before `config.shutdown()` clears telemetry, which the registration order guarantees.
- `getMetricsForSession`/`getSessionStartTimeForSession` are the right seam for the `/clear` case — the old global `getMetrics()`/`getSessionStartTime()` would have attributed pre- and post-clear usage to the same snapshot; the `/clear` handler flushes *before* the reset for exactly that reason.
- `flushSessionUsageSnapshot` is best-effort throughout (catch → false), so a telemetry failure can never block `/clear`, shutdown, or transcript deletion — consistent with the "usage reporting must never affect the session lifecycle" comment.
